bundles/af-core/src/main/java/com/adobe/cq/forms/core/components/internal/servlets/ReviewDataSourceServlet.java

```diff
@@ -0,0 +1,126 @@
+import com.adobe.cq.forms.core.components.models.form.Label;
+import com.adobe.cq.forms.core.components.util.ComponentUtils;
+import com.adobe.granite.ui.components.ExpressionResolver;
+import com.adobe.granite.ui.components.ds.DataSource;
+import com.adobe.granite.ui.components.ds.SimpleDataSource;
+import com.adobe.granite.ui.components.ds.ValueMapResource;
+
+@Component(
+    service = { Servlet.class },
+    property = {
+        "sling.servlet.resourceTypes=" + FormConstants.RT_FD_FORM_REVIEW_DATASOURCE_V1,
+        "sling.servlet.methods=GET",
+        "sling.servlet.extensions=html"
+    })
+public class ReviewDataSourceServlet extends AbstractDataSourceServlet {
+
+    /**
+     * Defines the form meta data type. Possible values: {@code submitAction},
+     * {@code prefillServiceProvider}
+     *
+     * @todo: Add other metadata types here like fragment, actions etc
+     */
+    @Reference
+    private transient ExpressionResolver expressionResolver;
+
+    @NotNull
+    @Override
+    protected ExpressionResolver getExpressionResolver() {
+        return expressionResolver;
+    }
+
+    @Override
+    protected void doGet(@NotNull SlingHttpServletRequest request, @NotNull SlingHttpServletResponse response) {
+
+        ResourceResolver resourceResolver = request.getResourceResolver();
+        String componentInstancePath = request.getRequestPathInfo().getSuffix();
+        List<Resource> resources = new ArrayList<>();
+        if (resourceResolver != null) {
+            Resource componentInstance = resourceResolver.getResource(componentInstancePath);
+            Resource formInstance = ComponentUtils.getFormContainer(componentInstance);
+            FormContainer formContainer = formInstance.adaptTo(FormContainer.class);
+            List<Base> panelList = (List<Base>) getPanels(formContainer);
+            List<Base> panels = panelList.stream().filter(x -> "panel".equals(x.getFieldType())).collect(Collectors.toList());
+            for (Base panel : panels) {
+                String name = panel != null ? panel.getName() : "";
+                String title = "";
+                if (panel != null) {
+                    Label label = panel.getLabel();
+                    if (label != null) {
+                        String value = label.getValue();
+                        if (value != null) {
+                            title = value;
+                        }
+                    }
+                }
+                if (name != null && title != null) {
+                    resources.add(getResourceForDropdownDisplay(resourceResolver, title, name));
+                }
+            }
+        }
+        SimpleDataSource actionTypeDataSource = new SimpleDataSource(resources.iterator());
+        request.setAttribute(DataSource.class.getName(), actionTypeDataSource);
+    }
+
+    private List<? extends ComponentExporter> getPanels(FormComponent formContainer) {
```

Comment:
It looks like this method doesn't just return Panels, it can return a field as well. The method name is misleading

Reply:
Have we tested this with a composite component like TnC?

Reply:
Changed, checked for all components